pkg/reconciler/workspaceindex/controller.go

```diff
@@ -0,0 +1,472 @@
+	// TODO: think through the concurrency here to make sure we're never getting extra events on credential reload
+	kcpSharedInformerFactory.Start(octx.Done())
+	kcpSharedInformerFactory.WaitForCacheSync(octx.Done())
+	klog.Infof("started informer for shard %s", workspace.Status.Location.Current)
+	return nil
+}
+
+func (c *Controller) enqueueWorkspace(shard, logicalCluster string, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	key = fmt.Sprintf("%s|%s|%s", shard, logicalCluster, key)
+	klog.Infof("queueing workspace %q", key)
+	c.workspaceQueue.Add(key)
+}
+
+func (c *Controller) startWorkspaceWorker(ctx context.Context) {
+	for c.processNextWorkspaceWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkspaceWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.workspaceQueue.Get()
+	if quit {
+		return false
+	}
+	c.activeWorkersLock.Lock()
+	c.workspaceQueueActiveWorkers += 1
+	c.activeWorkersLock.Unlock()
+	defer func() {
+		c.activeWorkersLock.Lock()
+		c.workspaceQueueActiveWorkers -= 1
+		c.activeWorkersLock.Unlock()
+	}()
+	key := k.(string)
+
+	klog.Infof("processing workspace %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.workspaceQueue.Done(key)
+
+	if err := c.processWorkspace(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.workspaceQueue.AddRateLimited(key)
+		return true
+	}
+	c.workspaceQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) processWorkspace(ctx context.Context, key string) error {
+	parts := strings.SplitN(key, "|", 3)
+	if len(parts) != 3 {
+		// programmer error
+		klog.Errorf("invalid key in workspace queue: %s", key)
+		return nil
+	}
+	var shard, logicalCluster string
+	shard, logicalCluster, key = parts[0], parts[1], parts[2]
+	var lister tenancylister.WorkspaceLister
+	if shard == rootShard {
+		lister = c.workspaceLister
+	} else {
+		informersByLogicalCluster, ok := c.shardInformers[shard]
+		if !ok {
+			klog.Infof("shard %s in workspace queue has no associated listers", shard)
+			return nil
+		}
+		informer, ok := informersByLogicalCluster[logicalCluster]
+		if !ok {
+			klog.Infof("shard %s in workspace queue has no associated lister for logical cluster %s", shard, logicalCluster)
+			return nil
+		}
+		lister = informer.workspaceLister
+	}
+	obj, err := lister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	obj = obj.DeepCopy()
+
+	return c.index.Record(obj)
+}
+
+// Stable exposes whether - to the best of our knowledge - the controller has stabilized.
+func (c *Controller) Stable() bool {
```

Comment:
am still not sure this is a viable approach. Stability of workspaces might never happen in every non-toy kcp.

Instead synchronously get the root org workspace once. Then start the org informers, synchronously wait for them to start up, but with some upper bound on waiting time (e.g. 60s). For every incoming org request (we know the org from the lcluster, right?) that is not in-sync return 50x with retry-after 10s or so.

Reply:
Maybe, we don't have to do that right away. Follow up is fine, but put a big TODO here.